File: apps/arc_cli/lib/arc/cli/apps.ex
defmodule Arc.CLI.Apps do
  @moduledoc """
  CLI commands for local ARC provider bundles.
  """

  alias Arc.CLI.ProviderBundle

  def run(args) do
    dispatch(args)
  end

  defp dispatch(["init"]) do
    init_bundle(".")
  end

  defp dispatch(["init", path | _]) do
    init_bundle(path)
  end

  defp dispatch(_) do
    IO.puts("""
    arc apps commands

    Commands:
      apps init [path]       Create a local ARC provider bundle scaffold
    """)
  end

  defp init_bundle(path) do
    case ProviderBundle.init(path) do
      {:ok, files} ->
        IO.puts("Initialized ARC app bundle at #{files.root}")
        IO.puts("")
        IO.puts("Created:")
        IO.puts("  #{files.arcfile}")
        IO.puts("  #{files.manifest}")
        IO.puts("  #{files.runtime}")
        IO.puts("")
        IO.puts("Next:")
        IO.puts("  ARC_KEY=<provider-key> bin/arc serve #{files.root}")
        IO.puts("  Edit manifest.json and run.sh to expose your capability")

      {:error, {:already_exists, path}} ->
        error("apps init failed: #{path} already exists")

      {:error, reason} ->
        error("apps init failed: #{inspect(reason)}")
    end
  end

  defp error(message) do
    IO.puts(:stderr, "error: #{message}")
    System.halt(1)
  end
end

File: go/cmd/arc/apps.go
package main

import (
	"fmt"

	"github.com/gezibash/arc/go/bundle"
	"github.com/spf13/cobra"
)

func appsCommand() *cobra.Command {
	command := &cobra.Command{
		Use:   "apps",
		Short: "Work with provider bundles",
	}

	command.AddCommand(&cobra.Command{
		Use:   "init [directory]",
		Short: "Write a new provider bundle",
		Long: "The bundle holds an Arcfile, a manifest, and a runtime that\n" +
			"answers one message. Serve it with arc serve <directory>.",
		Args: cobra.MaximumNArgs(1),
		RunE: appsInit,
	})
	return command
}

func appsInit(_ *cobra.Command, args []string) error {
	path := "."
	if len(args) == 1 {
		path = args[0]
	}

	files, err := bundle.Init(path)
	if err != nil {
		return err
	}

	fmt.Printf("wrote a bundle in %s\n", files.Root)
	fmt.Printf("  %s\n  %s\n  %s\n", files.Arcfile, files.Manifest, files.Runtime)
	fmt.Printf("\nserve it with: arc serve %s\n", path)
	return nil
}
